client/asset/btc/txdb.go

```diff
@@ -105,29 +113,75 @@ func NewBadgerTxDB(filePath string, log dex.Logger) (*BadgerTxDB, error) {
 		return nil, err
 	}
 
-	return &BadgerTxDB{
+	seq, err := db.GetSequence([]byte("seq"), 10)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, die := context.WithCancel(context.Background())
+
+	badgerDB := &BadgerTxDB{
 		DB:  db,
-		log: log}, nil
+		log: log,
+		seq: seq,
+		wg:  new(sync.WaitGroup),
+		die: die,
+	}
+	badgerDB.running.Store(true)
+
+	badgerDB.wg.Add(1)
+	go func() {
+		defer badgerDB.wg.Done()
+		ticker := time.NewTicker(5 * time.Minute)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				err := db.RunValueLogGC(0.5)
+				if err != nil && !errors.Is(err, badger.ErrNoRewrite) {
+					log.Errorf("garbage collection error: %v", err)
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return badgerDB, nil
+}
+
+// badgerDB returns ErrConflict when a read happening in a update (read/write)
+// transaction is stale. This function retries updates multiple times in
+// case of conflicts.
+func handleConflictWithBackoff(update func() error) error {
+	maxRetries := 10
+	sleepTime := 5 * time.Millisecond
+
+	var err error
+	for i := 0; i < maxRetries; i++ {
+		sleepTime *= 2
+		err = update()
+		if err != badger.ErrConflict {
+			return err
+		}
+		time.Sleep(10 * time.Millisecond)
```

Comment:
I'm not sure what you mean by passing in a context. I only want it to sleep if there is a conflict.

Reply:
We don't use `time.Sleep` unmonitored like this. If they cancel the app context, this function should also quit.

Reply:
Quitting in the middle of this function is kind of like quitting in the middle of a database transaction. The only reason this is sleeping is because the transaction previously conflicted with another transaction. I think we should let it retry instead of just dropping the transaction due to a conflict. This is not really "unmonitored ". In the new `Connect` function, when the context is cancelled, no new transaction will be allowed. Then it will wait for all transactions to complete before closing the database.